google/cloud/spanner/internal/time.cc

```diff
@@ -185,43 +192,92 @@ std::string TimestampToString(Timestamp ts) {
 StatusOr<Timestamp> TimestampFromString(std::string const& s) {
   std::tm tm;
   auto const len = s.size();
-  auto pos = ParseTime(kTimeFormat, s, &tm);
-  if (pos == std::string::npos || pos == len) {
+
+  // Parse full-date "T" time-hour ":" time-minute ":" time-second.
+  auto pos = ParseTime(s, &tm);
+  if (pos == std::string::npos) {
     return Status(StatusCode::kInvalidArgument,
                   s + ": Failed to match RFC3339 date-time");
   }
 
+  // Parse time-secfrac.
   Timestamp::duration ss(0);
-  if (s[pos] == '.') {
+  if (pos != len && s[pos] == '.') {
     Timestamp::duration::rep v = 0;
     auto scale = Timestamp::duration::period::den;
-    auto fpos = pos + 1;  // start of fractional part
+    auto fpos = pos + 1;  // start of fractional digits
     while (++pos != len) {
-      static constexpr auto kDigits = "0123456789";
       char const* dp = std::strchr(kDigits, s[pos]);
       if (dp == nullptr || *dp == '\0') break;  // non-digit
       if (scale == 1) continue;                 // drop insignificant digits
       scale /= 10;
       v *= 10;
-      v += dp - kDigits;
+      v += static_cast<Timestamp::duration::rep>(dp - kDigits);
     }
     if (pos == fpos) {
       return Status(StatusCode::kInvalidArgument,
                     s + ": RFC3339 time-secfrac must include a digit");
     }
-    ss = Timestamp::duration(v * scale);
+    ss = scale * Timestamp::duration(v);
   }
 
-  if (pos == len || s[pos] != 'Z') {
+  // Parse time-offset.
+  std::chrono::minutes utc_offset = std::chrono::minutes::zero();
+  bool offset_error = (pos == len);
+  if (!offset_error) {
+    switch (s[pos]) {
+      case 'Z':  // Zulu time
+      case 'z': {
+        ++pos;
+        break;
+      }
+      case '+':  // [-+]HH:MM
+      case '-': {
+        // Parse colon-separated hours, minutes, but not (yet) seconds.
+        std::array<std::chrono::minutes::rep, 2> fields = {{0, 0}};
+        auto it = fields.begin();
```

Comment:
> Should we refactor ParseInt() from time_format.cc and use it here?

I would say not.  The implementations are sufficiently simple and sufficiently distinct that trying to factor out commonality would, I think, be more trouble than it is worth.